app/services/spine_service/reports/clinic/diagnosis_report.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module SpineService
+  module Reports
+    module Clinic
+      class DiagnosisReport
+        include ModelUtils
+        attr_reader :start_date, :end_date
+
+        def initialize(start_date:, end_date:, **kwargs)
+          @start_date = start_date.to_date.beginning_of_day.strftime('%Y-%m-%d %H:%M:%S')
+          @end_date = end_date.to_date.end_of_day.strftime('%Y-%m-%d %H:%M:%S')
+        end
+
+        def fetch_report
+          type = EncounterType.find_by_name 'Outpatient diagnosis'
+          data = Encounter.where('encounter_datetime BETWEEN ? AND ?
+            AND encounter_type = ?
+            AND obs.concept_id IN(6543, 6542)', @start_date, @end_date, type.id).\
+            joins('INNER JOIN obs ON obs.encounter_id = encounter.encounter_id
+            INNER JOIN person p ON p.person_id = encounter.patient_id
+            LEFT JOIN person_name n ON n.person_id = encounter.patient_id AND n.voided = 0
+            LEFT JOIN person_attribute z ON z.person_id = encounter.patient_id AND z.person_attribute_type_id = 12
+            RIGHT JOIN person_address a ON a.person_id = encounter.patient_id
+            INNER JOIN concept_name c ON c.concept_id = obs.value_coded
+            ').\
+            group('obs.person_id,obs.value_coded,DATE(obs.obs_datetime)').\
+            select("encounter.encounter_type,n.given_name, n.family_name, n.person_id, obs.value_coded, p.gender,
+            a.state_province district, a.township_division ta, a.city_village village, z.value,
+            opd_disaggregated_age_group(p.birthdate,'#{end_date}') as age_group,c.name")
+
+            create_diagnosis_hash(data)
+        end
+
+        def create_diagnosis_hash(data)
```

Comment:
It will be great if you can define the report structure upfront. This will miss certain ages groups if the data or the period of the report does not have client within the age groups. For report consistency is best those indicators should not be missed at all. 

Most of these data manipulation can happen on the database level without having to do these on the application level. The age function should have a provision for spitting out unknown, the gender can be transformed right in the select clause